qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteCodeActions.java

```diff
@@ -298,4 +330,84 @@ private static void doCodeActionsForUndefinedSectionTag(Template template, Diagn
+
+	private static void doCodeActionsForSimilarValues(Template template, Diagnostic diagnostic,
+			List<CodeAction> codeActions, JavaDataModelCache javaCache, String varName) throws BadLocationException {
+		int offset = template.offsetAt(diagnostic.getRange().getStart());
+		Node node = QutePositionUtility.findBestNode(offset, template.findNodeAt(offset));
+
+		Range diagnosticRange = diagnostic.getRange();
+
+		Range rangeValue = new Range(
+				new Position(diagnosticRange.getStart().getLine(), diagnosticRange.getStart().getCharacter() + 1),
+				new Position(diagnosticRange.getEnd().getLine(), diagnosticRange.getEnd().getCharacter() - 1));
+
+		Collection<String> availableValues = collectAvailableValues(template, node, javaCache);
+		Collection<String> similarValues = new TreeSet<String>(Collator.getInstance());
+
+		for (String value : availableValues) {
+			if (isSimilar(value, varName)) {
+				similarValues.add(value);
+			}
+		}
+		if (!similarValues.isEmpty()) {
+			// Add code actions for each similar value
+			for (String similarValue : similarValues) {
+				CodeAction similarCodeAction = CodeActionFactory.replace("Did you mean '" + similarValue + "'?",
+						rangeValue, similarValue, template.getTextDocument(), diagnostic);
+				codeActions.add(similarCodeAction);
+			}
+		}
+	}
+
+	private static Collection<String> collectAvailableValues(Template template, Node node, JavaDataModelCache javaCache) {
+		Collection<String> availableValues = new TreeSet<String>(Collator.getInstance());
+
+		QuteProject project = template.getProject();
+		String projectUri = template.getProjectUri();
+
+		if (node.getKind() == NodeKind.ExpressionPart) {
+			Part previousPart = (((Part) node).getParent()).getPreviousPart((Part) node);
+			ResolvedJavaTypeInfo javaType = javaCache.resolveJavaType(previousPart, projectUri).getNow(null);
+			if (javaType != null) {
+				javaType.getMethods().stream().map(m -> availableValues.add(m.getMethodName()));
+			}
+		}
+
+		if (node.getKind() == NodeKind.Expression) {
+			Part objectPart = ((Expression) node).getObjectPart();
+
+			ResolvedJavaTypeInfo javaType = javaCache.resolveJavaType(objectPart, projectUri).getNow(null);
+			if (javaType != null) {
+				javaType.getMethods().stream().map(m -> availableValues.add(m.getMethodName()));
+			}
+
+			javaCache.getGlobalVariables(projectUri).stream().map(v -> availableValues.add(v.getName()));
+
+			List<String> aliases = template.getChildren().stream() //
+					.filter(n -> n.getKind() == NodeKind.ParameterDeclaration) //
+					.map(n -> ((ParameterDeclaration) n).getAlias()) //
+					.filter(alias -> alias != null) //
+					.collect(Collectors.toList());
+			aliases.stream().map(a -> availableValues.add(a));
+
+			ExtendedDataModelTemplate dataModel = javaCache.getDataModelTemplate(template).getNow(null);
+			if (dataModel != null && dataModel.getParameters() != null) {
+				dataModel.getParameters().stream().map(p -> availableValues.add(p.getKey()));
+			}
+		}

```

Comment:
`javaModelCache` works just fine. I can resolve various types from a sample project.

You need to debug this part because this is where the problems begin. You're passing `objectPart` to `javaCache.resolveJavaType(..)`, but why ? If it's unknown/undefined, what could it possibly return ?

Most of the stuff for `NodeKind.Expression` seems correct but it never gets triggered for me. Check what the node kinds are for the the undefined/unknown things you care about.

Reply:
Seem to have found the issue: the use of 
```java
javaCache.<getFromProjectRegistry>.stream().map(a -> availableValues.add(a));
```
wasn't populating the available values list as I thought it would.